### PR TITLE
[Router] Refactor PreloadLink and add global state pattern

### PR DESCRIPTION
**Pull request by damassi**
Cleans up / documents `<PreloadLink>` and lay out potential global state pattern via [`unstated`](https://github.com/jamiebuilds/unstated). 

Also adds transition hooks for tabs, which connects to PreloadLinkState: 

loader

### CODE REVIEW

**Comment by damassi** [comment]
This is a big PR but i'm gonna merge and keep going. We can iterate on this.

File: src/Router/PreloadLink.tsx [hunk 1]
```diff
@@ -1,177 +1,281 @@
...
+ *       return (
+ *         <PreloadLink to='/some-path'>
+ *           Fetching: {isFetching}
+ *         </PreloadLink>
+ *       )
+ *     }}
+ *   </Subscribe>
+ * )
+ */
+export class State extends Container<PreloadLinkState> {
```
Comment (damassi) [inline src/Router/PreloadLink.tsx]: This export pattern lets us take state that we're interested and "subscribe" to it anywhere in the tree. Keeps things nice and decoupled and great for lib code.

File: src/Styleguide/Pages/Artist/Components/LoadingArea.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        const loaderClass = isFetching ? "loading" : ""
+
+        return (
+          <React.Fragment>
+            <SpinnerToggle className={loaderClass} />
+            <Container className={loaderClass}>{props.children}</Container>
+          </React.Fragment>
+        )
+      }}
+    </Subscribe>
```
Comment (damassi) [inline src/Styleguide/Pages/Artist/Components/LoadingArea.tsx]: For example here. Import `PreloadLinkState` and subscribe, and now we've connected our main content area to the current state of the route request which enables some nice controlled behavior.

File: src/Styleguide/Pages/Artist/state.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      isLoggedIn: true,
+    })
+  }
+
+  logOut() {
+    this.setState({
+      isLoggedIn: false,
+    })
+  }
+}
```
Comment (damassi) [inline src/Styleguide/Pages/Artist/state.tsx]: Not being used ATM, just a sketch pad

Comment (alloy) [reply src/Styleguide/Pages/Artist/state.tsx]: How about we try to use local schema extension for this specific state?

File: src/Styleguide/Pages/Artist/Components/LoadingArea.tsx (hunk 2 above)
Comment (alloy) [reply src/Styleguide/Pages/Artist/Components/LoadingArea.tsx]: Doesn’t that tie this component to its parent? Couldn’t this be done as well by taking a `isLoading` prop and having the parent pass in its `isFetching` state?

File: src/Router/PreloadLink.tsx (hunk 1 above)
Comment (alloy) [reply src/Router/PreloadLink.tsx]: I feel like I’m missing something, the example being used in this change (and that I commented about below) is probably too contrived for me to see the bigger picture.

File: src/Styleguide/Pages/Artist/state.tsx (hunk 3 above)
Comment (damassi) [reply src/Styleguide/Pages/Artist/state.tsx]: Oh yeah, I'm curious about this, lets talk after meeting.

File: src/Styleguide/Pages/Artist/Components/LoadingArea.tsx (hunk 2 above)
Comment (damassi) [reply src/Styleguide/Pages/Artist/Components/LoadingArea.tsx]: The issue is that we're working across distant areas in the tree which make this sort of passing impractical and rigid. Will elaborate in Slack.

**Pull request merged**